### ISSUE TITLE
[RFC] Ascon Comportable IP Spec

### ISSUE THREAD

**Issue by m-temp** [lowRISC/opentitan#22179]
### TL;DR
This issue is to track  the [Google Doc Ascon Comportable IP Spec](https://docs.google.com/document/d/1hTnnhm5SoVCBYWObaTSfTLuKE07erR1ZapirLuR6YGU/edit?usp=sharing). The TC approved the RFC on March 21, 2023.

### Overview
This document specifies the [Ascon](https://ascon.iaik.tugraz.at/) hardware IP functionality. Ascon is a sponge-based, single pass, online Authenticated Encryption Scheme (that can also be used for hashing). It was selected by [NIST as the winner](https://www.nist.gov/news-events/news/2023/02/nist-selects-lightweight-cryptography-algorithms-protect-small-devices) of the NIST lightweight cryptography competition and is soon to be standardized.

### Features
The Ascon units supports the following features:
 - Authenticated encryption/decryption using either Ascon-128 (64 bit block size, 6 rounds per block update) or Ascon-128a (128 bit block size, 8 rounds per block update), whichever variant is standardized by NIST, with a key and tag size of 128 bit, which leads to 128 bit security, and a state size of 320 bits.
 - Support for Hashing might be implemented at a later time
 - First-order masking of the cipher core using domain-oriented masking (DOM) to deter side-channel analysis (SCA) is planned for the final version.
 - The latency of the Ascon permutation per 64/128 bit block is 6/8 clock cycles (unmasked) and presumably 12/16 clock cycles (DOM) for Ascon-128/Ascon-128a. Overall, 1 cycle is needed to absorb the data, 6/8 cycles are needed for the permutation and again 1 cycle to extract the output. However, the extraction and absorption are done simultaneously. Thus, in total, for long messages, the latency between a corresponding plaintext-ciphertext block is 1 cycle, and the latency between two consecutive ciphertext/plaintext blocks is 7/9 cycles; assuming large input sizes such that initialization and tag computation are negligible.
 - TBD. Automatic as well as software-initiated reseeding of internal pseudo-random number generators (PRNGs).
 - Countermeasures for deterring fault injection (FI) on the control path (for more details see [Security Hardening](https://docs.google.com/document/d/1hTnnhm5SoVCBYWObaTSfTLuKE07erR1ZapirLuR6YGU/edit#heading=h.96uzr5miw55)).
 - Register-based data and control interface.
 - System key manager interface for optional key sideload to not expose key material to the processor and other hosts attached to the system bus interconnect.

This Ascon unit targets medium performance (~1 clock cycle per round for the unmasked implementation). High-speed, multiple rounds per clock cycles for high-bandwidth data streaming is not required.

### ToDos

- [ ] File Skeleton PR
- [ ] Convert RFC into Markdown

**Comment by m-temp** [comment]
Duplicate https://github.com/lowRISC/opentitan/issues/22178
